File: string_matching/boyer_moore.cpp
#include "boyer_moore.h"
#include <array>
#include <algorithm>

namespace boyer_moore {
int search(const std::string &pattern, const std::string &text) {
    if (pattern.empty()) return 0;
    std::array<int, 256> bad;
    bad.fill(-1);
    for (size_t i = 0; i < pattern.size(); ++i) {
        bad[static_cast<unsigned char>(pattern[i])] = static_cast<int>(i);
    }
    size_t m = pattern.size();
    size_t n = text.size();
    size_t s = 0;
    while (s <= n - m) {
        int j = static_cast<int>(m - 1);
        while (j >= 0 && pattern[j] == text[s + j]) {
            --j;
        }
        if (j < 0) {
            return static_cast<int>(s);
        }
        unsigned char c = static_cast<unsigned char>(text[s + j]);
        int shift = std::max(1, j - bad[c]);
        s += shift;
    }
    return -1;
}
}

File: string_matching/boyer_moore.zig
const std = @import("std");

pub fn search(pattern: []const u8, text: []const u8) ?usize {
    if (pattern.len == 0) return 0;
    if (pattern.len > text.len) return null;
    var bad: [256]isize = undefined;
    for (bad) |*b| b.* = -1;
    var i: usize = 0;
    while (i < pattern.len) : (i += 1) {
        bad[pattern[i]] = @as(isize, @intCast(i));
    }
    var s: usize = 0;
    while (s + pattern.len <= text.len) {
        var j: isize = @as(isize, @intCast(pattern.len)) - 1;
        while (j >= 0 and pattern[@as(usize, @intCast(j))] == text[s + @as(usize, @intCast(j))]) {
            j -= 1;
        }
        if (j < 0) return s;
        const c = text[s + @as(usize, @intCast(j))];
        var shift: isize = j - bad[c];
        if (shift < 1) shift = 1;
        s += @as(usize, @intCast(shift));
    }
    return null;
}
